src/file/SDL_ioasync.c

```diff
@@ -0,0 +1,317 @@
+        return -1;
+    }
+    return ioasync->iface.size(ioasync->userdata);
+}
+
+static SDL_IOAsyncTask *RequestIOAsync(bool reading, SDL_IOAsync *ioasync, void *ptr, Uint64 offset, Uint64 size, SDL_IOAsyncGroup *group, void *userdata)
+{
+    if (!ioasync) {
+        SDL_InvalidParamError("ioasync");
+        return NULL;
+    } else if (!ptr) {
+        SDL_InvalidParamError("ptr");
+        return NULL;
+    } else if (!group) {
+        SDL_InvalidParamError("group");
+        return NULL;
+    }
+
+    SDL_IOAsyncTask *task = (SDL_IOAsyncTask *) SDL_calloc(1, sizeof (*task));
+    if (!task) {
+        return NULL;
+    }
+
+    task->ioasync = ioasync;
+    task->type = reading ? SDL_IOASYNC_TASK_READ : SDL_IOASYNC_TASK_WRITE;
+    task->offset = offset;
+    task->buffer = ptr;
+    task->requested_size = size;
+    task->app_userdata = userdata;
+    task->group = group;
+
+    SDL_LockMutex(ioasync->lock);
+    if (ioasync->closing) {
+        SDL_free(task);
+        SDL_UnlockMutex(ioasync->lock);
+        SDL_SetError("SDL_IOAsync is closing, can't start new tasks");
+        return NULL;
+    }
+    LINKED_LIST_PREPEND(task, ioasync->tasks, ioasync);
+    SDL_UnlockMutex(ioasync->lock);
+
+    const bool queued = reading ? ioasync->iface.read(ioasync->userdata, task) : ioasync->iface.write(ioasync->userdata, task);
+    if (!queued) {
+        SDL_LockMutex(ioasync->lock);
+        LINKED_LIST_UNLINK(task, ioasync);
+        SDL_UnlockMutex(ioasync->lock);
+        SDL_free(task);
+        task = NULL;
+    }
+
+    return task;
+}
+
+SDL_IOAsyncTask *SDL_ReadIOAsync(SDL_IOAsync *ioasync, void *ptr, Uint64 offset, Uint64 size, SDL_IOAsyncGroup *group, void *userdata)
+{
+    return RequestIOAsync(true, ioasync, ptr, offset, size, group, userdata);
+}
+
+SDL_IOAsyncTask *SDL_WriteIOAsync(SDL_IOAsync *ioasync, void *ptr, Uint64 offset, Uint64 size, SDL_IOAsyncGroup *group, void *userdata)
+{
+    return RequestIOAsync(false, ioasync, ptr, offset, size, group, userdata);
+}
+
+SDL_IOAsyncTask *SDL_CloseIOAsync(SDL_IOAsync *ioasync, SDL_IOAsyncGroup *group, void *userdata)
+{
+    if (!ioasync) {
+        SDL_InvalidParamError("ioasync");
+        return NULL;
+    } else if (!group) {
+        SDL_InvalidParamError("group");
+        return NULL;
+    }
+
+    SDL_LockMutex(ioasync->lock);
+    if (ioasync->closing) {
+        SDL_UnlockMutex(ioasync->lock);
+        SDL_SetError("Already closing");
+        return NULL;
+    }
+
+    SDL_IOAsyncTask *task = (SDL_IOAsyncTask *) SDL_calloc(1, sizeof (*task));
+    if (!task) {
+        SDL_UnlockMutex(ioasync->lock);
+        return NULL;
+    }
+
+    task->ioasync = ioasync;
+    task->type = SDL_IOASYNC_TASK_CLOSE;
+    task->app_userdata = userdata;
+    task->group = group;
+
+    ioasync->closing = task;
+
+    SDL_UnlockMutex(ioasync->lock);
+
+    if (LINKED_LIST_START(ioasync->tasks, ioasync) == NULL) { // no tasks? Queue the close task now.
```

Comment:
Doesn't the lock need to be held while examining and modifying ioasync->tasks?

Reply:
Whoops, yes, that's a bug.

Reply:
Fixed in 91d1484b59f04008712250ea5b142ae87e1fdff8.